Implement masking in FileType comparison on Unix

krtab opened the pull request:
Fixes: https://github.com/rust-lang/rust/issues/104900

krtab committed:
Implement masking in FileType comparison on Unix

Fixes: https://github.com/rust-lang/rust/issues/104900

the8472 commented:
Looks reasonable, but since the associated issue has a reproducer can you add it to the std testsuite? There's a helper function to get a temporary directory for the test, the dir creation should happen inside that.

krtab committed:
Add test for regression for FileType equality

Cf: https://github.com/rust-lang/rust/issues/104900

the8472 commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -332,11 +332,16 @@ pub struct FileTimes {
 }
 
-#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+#[derive(Copy, Clone, Eq, Hash, Debug)]
```
if we change equality behavior then we also need to change the `Hash` impl so that they're consistent.

krtab replied:
Oh, nice catch. But in this case, what would you think of rather implementing the invariant with a constructor? This would feel "cleaner" but the obvious disadvantage would be to have to add a module inside `unix:fs` to encapsulate `FileType`.

the8472 replied:
Constructor sounds good. A separate module isn't necessary since this is an internal type that isn't exported. It's not a soundness issue after all, only a correctness one.

krtab committed:
Implement masking in FileType hashing on Unix

Commit 77005950f09d2f9ba54962bf5adc3f2bc3a7213f implemented masking of
FileType to fix an issue[^1] in the semantic of FileType comparison.
This commit introduces masking to Hash to maintain the invariant that
x == y => hash(x) == hash(y).

[^1]: https://github.com/rust-lang/rust/issues/104900

the8472 commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -550,6 +562,10 @@ impl FileType {
     pub fn is(&self, mode: mode_t) -> bool {
         self.mode & libc::S_IFMT == mode
```
this can be changed to use `masked()` too

krtab committed:
Replace hand-made masking by call to masked() method in FileType

the8472 commented:
@bors r+

bors commented:
:pushpin: Commit 24cd863a3815f7ae7828c9f161a0b3f863218106 has been approved by `the8472`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=the8472 24cd863a3815f7ae7828c9f161a0b3f863218106 -->
<!-- homu: {"type":"Approved","sha":"24cd863a3815f7ae7828c9f161a0b3f863218106","approver":"the8472","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

krtab commented:
Thanks a lot for the review @the8472 !
